### PR TITLE
alphonse: round9 — Lion + surface_loss_weight=2.0 (binding-gap attack)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Test **surface_loss_weight=2.0** on the new EMA=0.999 SOTA stack — direct attack on the binding tau_y/tau_z and wall_shear gaps.

**Why surface_loss_weight=2.0:**

1. **Current weighting is unbalanced** — SOTA uses `--volume-loss-weight 2.0 --surface-loss-weight 1.0` (default). This gives volume loss 2× the gradient signal of surface loss. But our binding gaps are heavily on the surface side: tau_y ×3.7, tau_z ×3.9, wall_shear ×1.5, surface_pressure ×1.6 — vs volume_pressure ×2.1.
2. **Equal weighting may better balance** — boosting `--surface-loss-weight 2.0` matches the volume weight, making both loss components equally weighted. This should give surface predictions stronger gradient signal without dropping volume.
3. **Untested lever** — no surface_loss_weight sweep has been run. Volume_loss_weight=2.0 was inherited from yi calibration; surface_loss_weight stayed at default 1.0.
4. **PR #54 fern (per-axis tau weights on Lion) diverged** — but that touched per-component tau weighting, much riskier than the coarse surface vs volume balance.

**Risk:** volume_pressure may regress if model under-emphasizes volume. But we're not dropping volume_w — we're matching it, so total volume signal is preserved.

## Instructions

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer lion --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --surface-loss-weight 2.0 --volume-loss-weight 2.0 \
  --wandb-group tay-round9-lion-surface-w2
```

**Single delta vs SOTA PR #111:** adds `--surface-loss-weight 2.0` (was default 1.0). All other flags identical including EMA=0.999.

- Hard timeout 270 min (do NOT change SENPAI_TIMEOUT_MINUTES).
- Max 9 epochs — do not change SENPAI_MAX_EPOCHS.
- Post results in PR comment with: best val_primary/abupt_axis_mean_rel_l2_pct, final test_primary/abupt_axis_mean_rel_l2_pct, per-component test breakdown, W&B run URL.

## Baseline

Current SOTA — **PR #111 (tanjiro EMA=0.999, JUST MERGED), test_abupt 11.142**

| Metric | tay SOTA (PR #111) | AB-UPT | Gap |
|---|---:|---:|---:|
| `abupt` mean | **11.142** | — | — |
| `surface_pressure` | 6.209 | 3.82 | ×1.6 |
| `wall_shear` | 11.138 | 7.29 | ×1.5 |
| `volume_pressure` | 12.548 | 6.08 | ×2.1 |
| `tau_x` | 9.436 | 5.35 | ×1.8 |
| `tau_y` | 13.525 | 3.65 | **×3.7** |
| `tau_z` | 13.992 | 3.63 | **×3.9** |

The biggest binding gaps (tau_y ×3.7, tau_z ×3.9) are surface metrics. If surface_w=2.0 helps surface convergence, expect ≥3% improvement on tau_y/tau_z and surface_pressure.

W&B run: `ab3y4ej7` — 9 val epochs, best val 9.989 (ep9, first sub-10)

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT alphonse:

## Results

**NEW SOTA — surface_loss_weight=2.0 beats PR #111 by −4.99% (test_abupt 10.586 vs 11.142).** Surface gradient signal boost worked: 6 of 7 components improve, with the largest gains on the binding-gap targets (tau_y −5.93%, tau_z −7.18%, wall_shear −6.66%, surface_pressure −9.13%). Volume regressed slightly (+1.97%) but well within risk envelope.

### test_primary breakdown

| Metric | This run (surface_w=2.0) | PR #111 SOTA | AB-UPT ref | Δ vs SOTA |
|---|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.586** | 11.142 | — | **−4.99%** |
| `surface_pressure_rel_l2_pct` | **5.642** | 6.209 | 3.82 | **−9.13%** |
| `wall_shear_rel_l2_pct` | **10.396** | 11.138 | 7.29 | **−6.66%** |
| `volume_pressure_rel_l2_pct` | 12.795 | 12.548 | 6.08 | +1.97% |
| `wall_shear_x_rel_l2_pct` (tau_x) | **8.784** | 9.436 | 5.35 | **−6.91%** |
| `wall_shear_y_rel_l2_pct` (tau_y) | **12.723** | 13.525 | 3.65 | **−5.93%** |
| `wall_shear_z_rel_l2_pct` (tau_z) | **12.988** | 13.992 | 3.63 | **−7.18%** |

### val_primary per-epoch (EMA, full eval)

| Epoch | abupt | surface_p | wall_shear | vol_p | tau_x | tau_y | tau_z |
|---:|---:|---:|---:|---:|---:|---:|---:|
| 1 | 56.268 | 39.936 | 60.149 | 40.104 | 48.589 | 81.288 | 71.424 |
| 2 | 22.724 | 15.621 | 25.136 | 14.264 | 21.509 | 30.100 | 32.129 |
| 3 | 15.831 | 10.618 | 17.697 | 9.275 | 14.909 | 21.606 | 22.748 |
| 4 | 13.029 | 8.498 | 14.621 | 7.669 | 12.308 | 17.854 | 18.816 |
| 5 | 11.532 | 7.401 | 12.938 | 6.938 | 10.890 | 15.826 | 16.604 |
| 6 | 10.639 | 6.752 | 11.921 | 6.516 | 10.016 | 14.625 | 15.284 |
| 7 | 10.071 | 6.357 | 11.272 | 6.206 | 9.448 | 13.846 | 14.497 |
| 8 | 9.677 | 6.074 | 10.818 | 6.043 | 9.065 | 13.281 | 13.924 |
| **9** | **9.473** | **5.934** | **10.587** | **5.963** | **8.890** | **12.958** | **13.623** |

Best val ep9 = **9.473** (selected best-checkpoint, EMA source). PR #111 best val ep9 was 9.989 → −5.16% better.

### Run config

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --surface-loss-weight 2.0 --volume-loss-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-name "alphonse/round9-lion-surface-w2" \
  --wandb-group "tay-round9-lion-surface-w2"
```

- **W&B run (rank 0):** https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/okl62six
- **W&B group:** `tay-round9-lion-surface-w2` (8 ranks)
- **Wall-clock:** 270.8 min (DDP8)
- **Peak GPU memory:** ~74.2 GB allocated per rank (98.3% of 96 GB; reserved/peak across run hit ~101 GB momentarily during the run as reported by the wandb system metric)
- **Best epoch:** 9 (last epoch — still improving when budget cut in)
- **Best-checkpoint source:** EMA

### What happened

The hypothesis worked exactly as described. Boosting surface_loss_weight from 1.0 → 2.0 (matching the inherited volume_loss_weight=2.0) gives surface predictions equal gradient signal and pulls down all surface metrics meaningfully:

- **Largest gains exactly where predicted:** the PR called out tau_y (×3.7 gap) and tau_z (×3.9 gap) as the biggest binding gaps; both improved by ~6–7%. surface_pressure −9.13% is the biggest single-component delta.
- **Volume regression is mild and survivable:** +1.97% on volume_pressure is the cost, far smaller than the 4.99% mean gain. Total volume gradient signal is unchanged (volume_w stayed at 2.0); the regression is purely from the model's representational capacity reallocating slightly toward surface targets.
- **Per-epoch curves are still descending steeply at ep9** for all components — there's clear training-budget-bound improvement still on the table. Slopes (val/slope/...): abupt −0.122 pct/1k steps, tau_y −0.193, tau_z −0.180, wall_shear −0.138 — strongly negative across the board.
- **Train loss landed at 0.038**; grad norm pre-clip stable at ~1.23 throughout (clipping active at norm=1.0); zero NaN/inf grads. Stable run, no instability from the rebalance.

### Suggested follow-ups

1. **Longer budget for surface_w=2.0** — since slopes are still ~0.12 pct / 1k-steps at the cutoff, an extended training budget (e.g. 12–14 epochs) on this exact config could keep harvesting. The val curve hasn't plateaued.
2. **Push surface_w further (3.0, 4.0)** — the +1.97% volume regression is small; if a sweep at surface_w ∈ {2.5, 3.0, 4.0} keeps surface gains compounding while volume cost stays sub-linear, there may be more headroom. Diminishing returns expected at some point.
3. **Combine with per-component tau weighting** — PR #54 fern's per-axis tau weighting diverged on Lion, but with the new EMA=0.999 stack and the demonstrated robustness of surface_w=2.0, a gentler within-surface rebalance (e.g. tau_y/tau_z 1.5× tau_x/wall_shear, surface_pressure unchanged) might further close the binding gap on tau_y/tau_z (still ×3.5 vs AB-UPT).
4. **Try surface_w=2.0 + cosine schedule** — PR #110 (cosine T_max=50) gave −0.34% on the prior stack; combining cosine with this win is an obvious orthogonal lever.

The clean +5% improvement from a single hyperparameter and the per-component pattern matching the binding-gap thesis exactly suggests surface/volume balance was a meaningful unexplored axis. Happy to follow up on any of the above if the advisor wants.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #136 because ADVISOR: Closing — test_abupt 10.586 vs current SOTA 10.580 = **+0.06% (essentially tied, not a winner)**. 

The sw=2.0 lever IS real signal:
- surface_pressure −0.84% (5.642 vs 5.690)  
- wall_shear aggregate −0.22% (10.396 vs 10.419)
- tau_x −1.39% (8.784 vs 8.908)
- tau_z −0.64% (12.988 vs 13.071)

But it didn't help where intended (tau_y +1.86% to 12.723, the targeted binding gap WORSENED) and volume_pressure regressed by +0.43% (12.795 vs 12.740). The mixed-bag result means the lever is orthogonal but small.

**Important compound test in flight:** askeladd #141 is already running lr=1e-4 + sw=2.0 + EMA=0.999 (3-way compound). Result expected ~3hr. That experiment will tell us whether sw=2.0 stacks on the new SOTA stack or if the surface upweight is dominated by the lr lever.

**Reassigning alphonse** to PR #148: lr=1.5e-4 + EMA=0.999 on PR #115 SOTA stack. thorfinn's PR #115 results suggested testing higher LR (val curve still descending at ep9, gradient norms 0.5 with 0 clipping = headroom). Single delta from #115. If lr=1.5e-4 holds, lr=2e-4 next..
